paparazzi-gradle-plugin/src/main/java/app/cash/paparazzi/gradle/PaparazziPlugin.kt

```diff
@@ -51,21 +56,43 @@ import kotlin.io.path.invariantSeparatorsPathString
 class PaparazziPlugin : Plugin<Project> {
   override fun apply(project: Project) {
     project.afterEvaluate {
-      check(!project.plugins.hasPlugin("com.android.application")) {
-        error(
-          "Currently, Paparazzi only works in Android library -- not application -- modules. " +
-            "See https://github.com/cashapp/paparazzi/issues/107"
-        )
-      }
-      check(project.plugins.hasPlugin("com.android.library")) {
-        "The Android Gradle library plugin must be applied for Paparazzi to work properly."
+      val newResourceLoadingEnabled =
+        (project.findProperty("app.cash.paparazzi.new.resource.loading") as? String)?.toBoolean() ?: false
+
+      if (!newResourceLoadingEnabled) {
+        check(!project.plugins.hasPlugin("com.android.application")) {
+          error(
+            "Currently, Paparazzi only works in Android library -- not application -- modules. " +
+              "See https://github.com/cashapp/paparazzi/issues/107"
+          )
+        }
+
+        check(project.plugins.hasPlugin("com.android.library")) {
+          "The Android Gradle library plugin must be applied for Paparazzi to work properly."
+        }
+
+        project.plugins.withId("com.android.library") {
+          setupPaparazzi(project, project.extensions.getByType(LibraryExtension::class.java).libraryVariants)
+        }
+      } else {
+        check(listOf("com.android.application", "com.android.library", "com.android.dynamic-feature").any { project.plugins.hasPlugin(it) }) {
+          "The Android Gradle application or library or dynamic feature plugin must be applied for Paparazzi to work properly."
+        }
+
+        project.plugins.withId("com.android.application") {
+          setupPaparazzi(project, project.extensions.getByType(BaseAppModuleExtension::class.java).applicationVariants)
+        }
+        project.plugins.withId("com.android.dynamic-feature") {
+          setupPaparazzi(project, project.extensions.getByType(DynamicFeatureExtension::class.java).applicationVariants)
+        }
+        project.plugins.withId("com.android.library") {
+          setupPaparazzi(project, project.extensions.getByType(LibraryExtension::class.java).libraryVariants)
+        }
       }
     }
-
-    project.plugins.withId("com.android.library") { setupPaparazzi(project) }
   }
 
-  private fun setupPaparazzi(project: Project) {
+  private fun setupPaparazzi(project: Project, variants: DomainObjectSet<out BaseVariant>) {
     project.addTestDependency()
```

Comment:
`ApplicationVariant` and `LibraryVariant` both implement two interfaces `BaseVariant` and `TestedVariant`. I did not find a good generic type for the parameter, so have to pass as `BaseVariant` and then cast to `TestedVariant` later. Any suggestions?

Reply:
Generics: 

```diff
-private fun setupPaparazzi(project: Project, variants: DomainObjectSet<out BaseVariant>) {
+private fun <T> setupPaparazzi(project: Project, variants: DomainObjectSet<T>) where T : BaseVariant, T : TestedVariant {
```
and then you cam remove the cast (cast won't show as "unused" for some reason).